docs/proposals/blocks-storage-sharding.md

```diff
@@ -0,0 +1,84 @@
+3. Do not compromise correctness (either the query result is correct or it fails)
+
+## Proposed solution
+
+The idea is to introduce a new Cortex service - `store-gateway` - internally running the Thanos [`BucketStore`](https://github.com/thanos-io/thanos/blob/master/pkg/store/bucket.go). At query time, a querier will run a query fetching the matching series both from ingesters and the subset of gateways holding the related blocks (based on the query time range). Blocks are replicated across the gateways in order to guarantee query results consistency and HA even in the case of a gateway instance failure.
+
+### Ring-based sharding and replication
+
+In order to build blocks sharding and replication, the `store-gateway` instances form a [ring](../architecture.md#the-hash-ring). Each gateway instance uses a custom [`MetaFetcherFilter`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L108) to filter blocks loaded on the instance itself, keeping only blocks whose `hash(block-id)` is within the tokens range assigned to the gateway instance within the ring.
+
+Within a gateway, the blocks synchronization is triggered in two cases:
+
+1. **Periodically**<br />
+   to discover new blocks uploaded by ingesters or compactor, and delete old blocks removed due to retention or by the compactor
+2. **On-demand**<br/>
+   when the ring topology changes (the tokens ranges assigned to the gateway instance have changed)
+
+It's important to outline that the sync takes time (typically will have to re-scan the bucket and download new blocks index headers) and Cortex needs to guarantee query results consistency at any given time (_see below_).
+
+### Query execution
+
+When a querier executes a query, it will need to fetch series both from ingesters and the store-gateway instances.
+
+For a given query, the number of blocks to query is expected to be low, especially if the Cortex cluster is running the `query-frontend` with a `24h` query split interval. In this scenario, whatever is the client's query time range, the `query-frontend` will split the client's query into sub-queries each with up to `24h` time range and the querier will likely hit not more than 1 block per sub-query (except for the last 24h for which blocks may have not been compacted yet).
+
+Given this assumption, we want to avoid sending every query to every store-gateway instance. The querier should be able to take an informed decision about the minimum subset of store-gateway instances which needs to query given a time range.
+
+The idea is to run the [`MetaFetcher`](https://github.com/thanos-io/thanos/blob/master/pkg/block/fetcher.go#L127) also within the querier, but without any sharding filter (contrary to the store-gateway). At any given point in time, the querier knows the entire list of blocks in the storage. When the querier executes the `Select()` (or `SelectSorted()`) it does:
```

Comment:
Nice! Interesting idea :+1: 

cc @brancz as we already talked about block-aware querying for Querier. I wish we could do it on Thanos, but there it's not that easy - not all StoreAPI are block based, so our Querier does not have this knowledge.